### ISSUE TITLE
Database update fails with KeyError (caused by Inspect.Error)

### ISSUE THREAD

**Issue by jec** [dominique-vassard/seraph#12]
I can't get a Seraph node to insert into Neo4j. Here is my setup.

```
$ elixir -v
Erlang/OTP 23 [erts-11.1.8] [source] [64-bit] [smp:12:12] [ds:12:12:10] [async-threads:1] [hipe]

Elixir 1.11.3 (compiled with Erlang/OTP 23)
```

_mix.exs:_
```elixir
#...
  defp deps do
    [
      #...
      {:phoenix, "~> 1.5.8"},
      {:seraph, "~> 0.2"},
      #...
    ]
  end
#...
    
```

_config/test.exs:_
```elixir
config :my_app, MyApp.Repo,
       url: "bolt://localhost:7687",
       basic_auth: [username: "neo4j", password: "****"],
       pool_size: 10
```

_lib/my_app/application.ex:_
```elixir
  def start(_type, _args) do
    children = [
      #...
      MyApp.Repo,
      #...
    ]
```

_lib/my_app/repo.ex:_
```elixir
defmodule MyApp.Repo do
  use Seraph.Repo, otp_app: :my_app
end
```

_lib/my_app/session/user.ex:_
```elixir
defmodule MyApp.Session.User do
  use Seraph.Schema.Node
  import Seraph.Changeset

  node "User" do
    property :identifier, :string
    property :password, :string, redact: true
    property :name, :string
    property :email, :string
  end

  def changeset(%MyApp.Session.User{} = user, params \\ %{}) do
    user
    |> cast(params, [:identifier, :password, :name, :email])
  end
end
```

_test/my_app/session_test.exs:_
```elixir:
defmodule MyApp.SessionTest do
  use MyApp.Case

  alias MyApp.Session

  describe "Session.log_in/2" do
    test "with valid credentials and user, returns status 200" do
      user = %Session.User{identifier: "phred"}
      IO.inspect(user)
      {:ok, user} = MyApp.Repo.Node.create(user)
    end
  end
end
```

When I run `mix test test/my_app/session_test.exs`, it outputs:

```
%MyApp.Session.User{
  __id__: nil,
  __meta__: %Seraph.Schema.Node.Metadata{
    primary_label: "User",
    schema: MyApp.Session.User
  },
  additionalLabels: [],
  email: nil,
  identifier: "phred",
  name: nil,
  password: nil,
  uuid: nil
}


  1) test Session.log_in/2 with valid credentials and user, returns status 200 (MyApp.SessionTest)
     test/my_app/session_test.exs:5
     ** (KeyError) key :repo_changes not found in: %Inspect.Error{message: "got FunctionClauseError with message \"no function clause matching in MyApp.Session.User.__schema__/1\" while inspecting %{__struct__: Ecto.Changeset, action: nil, changes: %{additionalLabels: [], identifier: \"phred\"}, constraints: [], data: %MyApp.Session.User{__id__: nil, __meta__: %Seraph.Schema.Node.Metadata{primary_label: \"User\", schema: MyApp.Session.User}, additionalLabels: nil, email: nil, identifier: nil, name: nil, password: nil, uuid: nil}, empty_values: [\"\"], errors: [], filters: %{}, params: %{\"additionalLabels\" => [], \"email\" => nil, \"identifier\" => \"phred\", \"name\" => nil, \"password\" => nil, \"uuid\" => nil}, prepare: [], repo: nil, repo_opts: [], required: [], types: nil, valid?: false, validations: []}"}. Did you mean one of:
     
           * :changes
           * :repo
     
     code: {:ok, user} = MyApp.Repo.Node.create(user)
     stacktrace:
       (stdlib 3.14) :maps.update(:repo_changes, %{}, %Inspect.Error{message: "got FunctionClauseError with message \"no function clause matching in MyApp.Session.User.__schema__/1\" while inspecting %{__struct__: Ecto.Changeset, action: nil, changes: %{additionalLabels: [], identifier: \"phred\"}, constraints: [], data: %MyApp.Session.User{__id__: nil, __meta__: %Seraph.Schema.Node.Metadata{primary_label: \"User\", schema: MyApp.Session.User}, additionalLabels: nil, email: nil, identifier: nil, name: nil, password: nil, uuid: nil}, empty_values: [\"\"], errors: [], filters: %{}, params: %{\"additionalLabels\" => [], \"email\" => nil, \"identifier\" => \"phred\", \"name\" => nil, \"password\" => nil, \"uuid\" => nil}, prepare: [], repo: nil, repo_opts: [], required: [], types: nil, valid?: false, validations: []}"})
       (ecto 3.5.8) lib/ecto/changeset.ex:278: anonymous fn/2 in Ecto.Changeset.__struct__/1
       (stdlib 3.14) maps.erl:233: :maps.fold_1/3
       (elixir 1.11.3) lib/kernel.ex:2214: Kernel.struct!/2
       (seraph 0.2.2) lib/seraph/changeset.ex:161: Seraph.Changeset.apply_changes/1
       (seraph 0.2.2) lib/seraph/repo/node/schema.ex:16: Seraph.Repo.Node.Schema.create/3
       test/my_app/session_test.exs:8: (test)



Finished in 0.02 seconds
1 test, 1 failure

Randomized with seed 822369
```

It's failing on the `struct!` call in `Seraph.Changeset.map_to_ecto/1`:

```elixir
  defp map_to_ecto(changeset) do
    struct!(Ecto.Changeset, Map.from_struct(changeset))
  end
```

I'm not sure what to try from here. I've looked at my setup and compared it to the Seraph "Up and Running" page multiple times, but I can't find anything that I've done wrong. Maybe I've just been looking at it too much.

Thanks for any help you can provide.

**Comment by jec** [comment]
Because of the `no function clause matching in MyApp.Session.User.__schema__/1` error, I added a catch-all for `__schema__/1` to `Seraph.Schema.Node`. The unmatched argument is `:redact_fields`.

**Comment by jec** [comment]
I added the following to my _lib/my_app/session/user.ex:_

```elixir
  def __schema__(:redact_fields), do: [:password]
```

and now the test error is:

```
  1) test Session.log_in/2 with valid credentials and user, returns status 200 (MyApp.SessionTest)
     test/my_app/session_test.exs:5
     ** (KeyError) key :repo_changes not found in: #Ecto.Changeset<action: nil, changes: %{additionalLabels: [], identifier: "phred"}, errors: [], data: #MyApp.Session.User<>, valid?: false>. Did you mean one of:
     
           * :changes
           * :repo
     
     code: {:ok, user} = MyApp.Repo.Node.create(user)
     stacktrace:
       (stdlib 3.14) :maps.update(:repo_changes, %{}, #Ecto.Changeset<action: nil, changes: %{additionalLabels: [], identifier: "phred"}, errors: [], data: #MyApp.Session.User<>, valid?: false>)
       (ecto 3.5.8) lib/ecto/changeset.ex:278: anonymous fn/2 in Ecto.Changeset.__struct__/1
       (stdlib 3.14) maps.erl:233: :maps.fold_1/3
       (elixir 1.11.3) lib/kernel.ex:2214: Kernel.struct!/2
       (seraph 0.2.2) lib/seraph/changeset.ex:161: Seraph.Changeset.apply_changes/1
       (seraph 0.2.2) lib/seraph/repo/node/schema.ex:16: Seraph.Repo.Node.Schema.create/3
       test/my_app/session_test.exs:8: (test)
```

**Comment by dominique-vassard** [comment]
Hi,
Thanks for raising the issue.
It's due to some conflicts with ecto version. 
I'll push a fix shortly.

**Comment by jec** [comment]
I've created a repo at https://github.com/jec/my_app with a minimal app that demonstrates this issue. Interestingly, it doesn't require the `def __schema__(:redact_fields), do: [:password]` in the User schema to reveal the actual error:

```
..%MyApp.Auth.User{
  __id__: nil,
  __meta__: %Seraph.Schema.Node.Metadata{
    primary_label: "User",
    schema: MyApp.Auth.User
  },
  additionalLabels: [],
  email: "phred@example.net",
  identifier: "phred",
  name: "Phred",
  password: "foobar",
  uuid: nil
}


  1) test can create a User node (MyApp.Auth.UserTest)
     test/my_app/auth/user_test.exs:4
     ** (KeyError) key :repo_changes not found in: #Ecto.Changeset<action: nil, changes: %{additionalLabels: [], email: "phred@example.net", identifier: "phred", name: "Phred", password: "foobar"}, errors: [], data: #MyApp.Auth.User<>, valid?: false>. Did you mean one of:
     
           * :changes
           * :repo
     
     code: {:ok, user} = MyApp.Repo.Node.create(user)
     stacktrace:
       (stdlib 3.14) :maps.update(:repo_changes, %{}, #Ecto.Changeset<action: nil, changes: %{additionalLabels: [], email: "phred@example.net", identifier: "phred", name: "Phred", password: "foobar"}, errors: [], data: #MyApp.Auth.User<>, valid?: false>)
       (ecto 3.2.5) lib/ecto/changeset.ex:274: anonymous fn/2 in Ecto.Changeset.__struct__/1
       (stdlib 3.14) maps.erl:233: :maps.fold_1/3
       (elixir 1.11.3) lib/kernel.ex:2214: Kernel.struct!/2
       (seraph 0.2.2) lib/seraph/changeset.ex:161: Seraph.Changeset.apply_changes/1
       (seraph 0.2.2) lib/seraph/repo/node/schema.ex:16: Seraph.Repo.Node.Schema.create/3
       test/my_app/auth/user_test.exs:12: (test)
```

**Comment by dominique-vassard** [comment]
Hi,
issue is solved, and version 0.2.3 is published.

Thanks again for the issue and the neat repository!

**Comment by jec** [comment]
Node.create is working now. Thanks!
